Added features

dhskinner opened the pull request:
Hi there - I had a play with the code throughout today and hope its not too forward but did a bit of a re-write to add some features and flexibility. Changes look like a lot but they're really just building on the existing codebase (and huge thank you Phil for all your hard work on this!). Testing was limited to a setup I have with a couple of NMEA2K sensors so suggestion is to fork the code and test it a little before adding to the main branch as a new release if you're ok with this.
Summary of changes:
- now able to push numeric, text and boolean data types to InfluxDB-V2, and other data types as JSON
- added ability to push data from contexts outside of 'vessel.self'
- added ability to expand properties of any measurement i.e. building on previous ability
to expand position and attitude - now any measurement with multiple properties can be expanded
- added ability to add tags against each individual path
- added the context & path to the name of each measurement
- improved handling of wildcard '*' for context and path
- lots of unnecessary refactoring and tidy-up
Please do take it (or leave it...) and let me know if anything doesn't work! Dave

dhskinner committed:
Update index.js

Add new functionality:
    -   now able to push numeric, text and boolean data types to InfluxDB-V2, and other data types as JSON
    -   added ability to push data from contexts outside of 'vessel.self'
    -   added ability to expand properties of any measurement (i.e. building on previous ability
         to expand position and attitude, now any measurement with multiple properties can be expanded)
    -   added ability to add tags against each individual path
    -   added the context & path to the name of each measurement 
    -   improved handling of wildcard '*' for context and path 
    -   lots of unnecessary refactoring and tidy-up

dhskinner committed:
Merge pull request #1 from plainolddave/plainolddave-patch-1

Update index.js

pbegg commented:
Thanks plainolddave! I haven't been giving this plugin as much love as I should. I appreciate the pull request, I'll have a look over it in the next week or so. I've also made some changes and not updated. Mainly around buffering data with drop of internet connection

dhskinner commented:
No worries thanks Phil

Apologies in advance it looks like a complete re-write but its not really – just moved a bunch of stuff around / am not precious so feel free to use or discard it as you like 😊

Unfortunately I kind of ran out of time today to solve the thing I really was hoping to get on top of which is to push vessel data from AIS to a map in Grafana – still trying to understand how to link all the separate signalK measurements together so vessel name, position and whatnot all go to influx together….. 

Appreciate your thoughts and happy to be corrected

Dave

dhskinner committed:
Update index.js

Improved tagging, and added auto tags for measurements from 'self'

dhskinner committed:
Revert "Update index.js"

This reverts commit 45fb3fca23fe7b83407f3766b9ede911367d9812.

dhskinner committed:
Improved tagging

Improved tags and bug-fixes. Allows easier filtering in Influx by including the measurement context,  source and flag for 'self'

dhskinner commented:
Hi again - I made a couple of tweaks a fortnight ago and it all seems stable. The only thing is the user interface for the plugin config screen is a little unwieldy - I can't for the life of me figure out how UISchema works to put some collapsible menus in place so its a little more pleasing to the eye. Let me know thoughts / thanks again!

pbegg commented:
Hey,

Sorry haven’t had a chance to look at it yet. Been a hectic few weeks. Will
try to look soon.
Glad it’s seems stable. Have you implemented anything for buffering without
an internet connection?

On Fri, 9 Sep 2022 at 4:12 pm, plainolddave ***@***.***>
wrote:

> Hi again - I made a couple of tweaks a fortnight ago and it all seems
> stable. The only thing is the user interface for the plugin config screen
> is a little unwieldy - I can't for the life of me figure out how UISchema
> works to put some collapsible menus in place so its a little more pleasing
> to the eye. Let me know thoughts / thanks again!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pbegg/signalk-to-influxdb-v2-buffering/pull/19#issuecomment-1241540020>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMSRLZX5PZH7O6IN6FTXVHDV5LIMFANCNFSM57HBIS2A>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

mhaberler commented:
@plainolddave this looks great, I guess I will give it a spin!

pbegg commented:
My plan is to give this a run over the weekend. 
Looks really good - I like that you have included the source and can expand the mapped objects.

dhskinner committed:
Added a checkbox to enable or disable paths

Added a checkbox to enable or disable paths

dhskinner commented:
> Have you implemented anything for buffering without an internet connection?

Good question... have been thinking about this... for myself am running SignalK on a tiny setup in a Raspberry Pi so buffering in RAM is limited (and continuous read/writes of data to an SD card is fraught) so its not something I've needed or tried.

Am thinking maybe the answer is to push data to a local instance of Influx OSS that replicates itself when connectivity returns using edge-data-replication?  Like this [https://docs.influxdata.com/influxdb/cloud/write-data/replication/](https://docs.influxdata.com/influxdb/cloud/write-data/replication/). This could give far bigger capacity (i.e. just limited by hard drive space) for folks out on the water for weeks on end like cruising yachties....

Setups like mine on a RasPi would need to mount an external SSD or small NAS to avoid wear on the SD card. There's another boat due for a refit in a month or two, so buffering/replication is on the list of things to play with... just a thought here

mgf909 commented:
HI guys - I have been following this plugin and daves changes recently and have just setup a new system ( new boat ) so im running Dave's version on a small pc on my boat. I installed the original version via the SK "appstore" then replaced the updated index.js into my system.. I configured it use local install of Influxdb2 ( http://127.0.0.1:8086 ) - all seems to work fine and ive knocked up a basic Grafana dash ( also running locally ). Im yet to give it a long run yet however.... please let me know if i can help test etc. 

Slightly off topic, although relevant to the plugin...
How does this handle text strings?  ive not enabled logging of "notifications.*" just yet...should it work?
My engine gateways send these messages - would be great to write these to the database : <timestamp>, <message>,<tag>,<tag> or similar so these could also be stored..
`
{   "state": "normal",   "method": [     "visual"   ],   "message": "Port Engine Charge Indicator is Normal" }
`

ump-onwatch commented:
@plainolddave Finally got around to testing this - looks great.

dhskinner commented:
Hi MGF909 - sorry for delayed response - regular text like _"state": "normal"_ is fine - there's no issues handling similar strings like vessel names from AIS on my small test setup

Can you let us know if the _"method": [ "visual" ]_ value comes through ok? It should in theory just be sent as a JSON array but  am curious if the code needs a tweak...

Thanks

Dave

pbegg commented:
Hey Dave,

Ran a test with some engine data coming through and producing notifications. This is is the output on Influx:


<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/phil/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/phil/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
-->

</head>

<body link="#0563C1" vlink="#954F72">



_start | _stop | _time | _value | _field | _measurement
-- | -- | -- | -- | -- | --
2022-10-13T06:53:30.298717449Z | 2022-10-13T06:54:30.298717449Z | 2022-10-13T06:53:31.019Z | ["visual"] | value | notifications.propulsion.port.preheatIndicator.method



</body>

</html>


Is this the intended outcome?

It looks great appreciate the clean up you have done!

What are you thoughts on the 'notification' path  having a seperate format:

- the notification path should be measurement.
- State should be the value
- method and message could be tags?

mhaberler commented:
good point - with reporting one measurement and timestamp per path, it becomes difficult to correlate several measurements by time

I'd suggest a generic measurement name and one field per value, plus maybe tags for events

like measurment=navigation, lat=47,lon=15  event='3d fix'

dhskinner commented:
Hi [Phil](https://github.com/pbegg) - yep good point

Agree it probably needs more thought as there's three fields required by the influxdb-client 'Point' object imported from "Point.d.ts" -> _measurement (that's defined in the Point constructor), _field, and  _value. These can be populated however you like - currently they map as:
- _measurement = signalK path
- _field = is currently used to expand multiple properties like lat/lon (more on this below). If the measurement only has one property then its hard-coded to 'value'
- _value = signalK data measurement

The  limitation I struck is adding extra columns for multiple data values in each row like as suggested by [mhaberler](https://github.com/mhaberler)  quickly seemed to hit the cardinality limits of InfluxDB Cloud - the reason this came up is I was playing with AIS data, where SignalK gives a new context to each vessel, and Influx started complaining (right now there's 137 vessels on AIS around me so each combo of MMSI & position, speed, course drives up cardinality). Shouldn't really matter though if you're just picking up data from 'vessel.self'

Might need to experiment more here as to be honest I don't 'get' what InfluxDB Cloud expects to see when trying to add multiple values in a single row - right now everything other than the '_value' column all seems to just be ingested as tags  (hope this makes sense). 

The workaround is to either split values with multiple properties (e.g. position) into different rows (lat, lon) and recombine them with FluxQL in Grafana, or to leave the properties encoded to one row as JSON which can then also be fairly easily unpacked in Grafana (you can see the workaround either way is somehow get the data into influx and deal with it later in grafana)

On that point - on re-looking at the code I've just realised the realised the 'Expand properties' needed a tweak in a(nother) commit so it works like this:
- when 'Expand properties' is not selected it sends measurements like position as JSON:
<html>
<body>
<!--StartFragment-->

table | _measurement | _field | _value | _time | 
-- | -- | -- | -- | -- 
0 | navigation.position | value | {"longitude":153.216955,"latitude":-27.700000} | 2022-10-23T04:41:47.796Z

<!--EndFragment-->
</body>
</html>
- when 'Expand properties' is selected it sends measurements like position as multiple rows:
<html>
<body>
<!--StartFragment-->

table | _measurement | _field | _value | _time | 
-- | -- | -- | -- | -- 
0 | navigation.position | latitude | -27.700000 | 2022-10-23T04:44:42.080Z
1 | navigation.position | longitude | 153.216955 | 2022-10-23T04:44:42.080Z

<!--EndFragment-->
</body>
</html>

Hope this helps - let me know thoughts 

Dave

dhskinner committed:
Update index.js

dhskinner commented:
Hi again - one other thought Phil - did you want to grab the code from the latest commit and just put it in another branch, that way yourself (or anyone) can play around to get it sorted?

Whatever is easiest, but may be better than waiting for me to occasionally read email and update this pull request :) 

Thanks again 

Dave

mhaberler commented:
@plainolddave I can imagine unrolling compound objects like position into separate measurements 

when querying such an unrolled object and effectively re-joining into a compound object, the components better have the same timestamp

so that must be guaranteed by setting the same timestamp on all components at injection time

btw - what is the take on timestamp resolution for this use case?
seconds? milliseconds?

dhskinner commented:
Hi @mhaberler 

> when querying such an unrolled object and effectively re-joining into a compound object, the components better have the same timestamp so that must be guaranteed by setting the same timestamp on all components at injection time

Yes - values from the same update in signalK have identical timestamp (e.g. position -> lat and lon)

> btw - what is the take on timestamp resolution for this use case?  seconds? milliseconds?

Resolution is milliseconds e.g. 2022-10-23T04:44:42 .080 Z

Thanks

pbegg commented:
Pulled to new branch - v3-newfeatures
